HLSL fixed function implementation

NZJenkins opened the pull request:
Should behave more-or-less the same as D3D9 fixed function mode
But allow us to handle Xbox extensions and idiosyncrasies better e.g. backface lighting

F2 toggles back to D3D9 fixed function rendering

Co-authored with @PatrickvL

Opening as draft in case anyone feels like looking for regressions. Everything should generally look exactly the same

PatrickvL commented on projects/cxbx/CMakeLists.txt:
```diff
@@ -189,3 +189,18 @@ if(${CMAKE_GENERATOR} MATCHES "Visual Studio ([^9]|[9][0-9])")
+# Copy shader files to hlsl directory
+add_custom_target(copy_hlsl ALL
+ COMMAND ${CMAKE_COMMAND} -E make_directory $<TARGET_FILE_DIR:cxbx>/hlsl
+ COMMAND ${CMAKE_COMMAND} -E copy "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShaderState.cpp" $<TARGET_FILE_DIR:cxbx>/hlsl
```
maybe change extension into `.hlsli` ?

NZJenkins replied:
makes sense, will do

PatrickvL commented on src/core/hle/D3D8/Direct3D9/Direct3D9.cpp:
```diff
@@ -6314,6 +6322,142 @@ void CreateHostResource(xbox::X_D3DResource *pResource, DWORD D3DUsage, int iTex
+	// Write fixed function state to shader constants
+	const int slotSize = 16;
+	const int fixedFunctionStateSize = (sizeof(FixedFunctionVertexShaderState) + slotSize - 1) / slotSize;
+	auto hRet = g_pD3DDevice->SetVertexShaderConstantF(0, (float*)&ffShaderState, fixedFunctionStateSize);
```
Log when `!SUCCEEDED(hRet)`?

PatrickvL commented on src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShader.hlsl:
```diff
@@ -0,0 +1,600 @@
+    if(index == normal) return xIn.normal;
+    if(index == diffuse) return xIn.color[0];
+    if(index == specular) return xIn.color[1];
+    //if(i == fogCoord) xIn.fog = value;
```
Although commented out (why?), best `return` a value

NZJenkins replied:
vaguely remember Xbox FVF didn't take FOG or PSIZE (but host D3D9 FVF supports PSIZE?)
will check against the vertex pr and see what we were doing

PatrickvL replied:
Okay, so then it should probably read : `assert(i != fogCoord); // fogCoord is never part of any Xbox FVF` and similar for `pointSize`. (Not sure how to implement `assert`, but even having it stubbed would still improve clarity.)

NZJenkins replied:
just removed this since its not part of the inputs.
the reason being its not used in DoFog - since the fog value is stored in the specular alpha

PatrickvL commented on src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShader.hlsl:
```diff
@@ -0,0 +1,600 @@
+        float3 halfway = normalize(toViewerVN + toLightVN);
+        float NdotH = dot(View.Normal, halfway);
+
+        float3 frontSpecular = pow(abs(NdotH), powers[0]) * lightSpecular.rgb;
+        float3 backSpecular = pow(abs(NdotH), powers[1]) * lightSpecular.rgb;
```
All `powers` variables passed along are used only here, for specular light calculation. Therefore, best rename it into `specular_powers` everywhere.

NZJenkins replied:
True - wonder if it's worth bundling `toViewerVN`, `powers`, and `lightSpecular` into a `SpecularInfo` or something
I think they're all specular specific

NZJenkins replied:
lighting calculations including DoSpecular were consolidated into one method, so variables aren't needlessly passed round any more

PatrickvL commented on src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShader.hlsl:
```diff
@@ -0,0 +1,600 @@
+			texCoord = View.Position;
+		else if (tState.TexCoordIndexGen == TCI_CAMERASPACEREFLECTIONVECTOR)
+			texCoord.xyz = reflected;
+        // else if TCI_OBJECT TODO is this just model position?
```
How could unhandled cases like this be detected? Make it somehow visually stand out? But how?

NZJenkins replied:
In this particular case it should be apparent since it's texture coordinates

NZJenkins replied:
Added log test cases when the state is set. Previously we logged

PatrickvL commented on src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShader.hlsl:
```diff
@@ -0,0 +1,600 @@
+	        if(i == normal) xIn.normal = value;
+	        if(i == diffuse) xIn.color[0] = value;
+	        if(i == specular) xIn.color[1] = value;
+	        //if(i == fogCoord) xIn.fog = value;
+	        //if(i == pointSize) xIn.psize = value;
```
Why are these commented out, again?

NZJenkins replied:
unused and not part of VS_INPUT

PatrickvL commented on src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShader.hlsl:
```diff
@@ -0,0 +1,600 @@
+        #else
+	        // switch statements inexplicably don't work here
+	        if(i == position) xIn.pos = value;
+	        if(i == weight) xIn.bw = value;
```
Are the else statements left out for efficiency? Perhaps explain in a comment.

NZJenkins replied:
Hit an issue today where `if` and `if else` can be the difference between the shader compiling and `X4505: maximum temp register index exceeded` (in unrelated parts of the shader!)
So maybe the compiler doesn't like `else`

Anyway, nothing to explain here, other than `switch` didn't seem to work at the time
You would think, branching on a single variable, there would be no difference between `if`, `if else`, and `switch` 🤔

PatrickvL commented on src/core/hle/D3D8/FixedFunctionState.cpp:
```diff
@@ -0,0 +1,84 @@
+
+    // Replace the oldest element and move to end
+    EnabledLights[0] = index;
+    std::rotate(std::begin(EnabledLights), std::begin(EnabledLights) + 1, std::end(EnabledLights));
```
Why the shuffling around of lights? If it's a requirement so that the fixed function shader can access just the enabled lights in a linear fashion, then that rationally should probably be explained in a comment somewhere.

NZJenkins replied:
Does the comment in the header file help? I should add some explanation here too
We are holding onto the 8 most recently enabled lights. So if a title activates 9 lights we can disable the oldest one (at index 0)
IIRC that's what D3D9 does anyway, not sure of a test case though.

Oops, I think the enable case rotates the wrong way!

PatrickvL replied:
The comment in the header states what it does, but not why it's required to be that way

PatrickvL commented on src/core/hle/D3D8/XbVertexShader.h:
```diff
@@ -107,6 +107,18 @@ extern D3DVERTEXELEMENT *EmuRecompileVshDeclaration
 );
 
+enum class VertexShaderMode {
+	FixedFunction,
+	// TODO implement the FixedFunction vs. Passthrough distinction
+	// When titles use Xbox fixed function with pre-transformed vertices
+	// it actually uses a special "passthrough" shader program
+	// Passthrough,
+	ShaderProgram
```
To avoid symbol conflicts, it may be good to prefix these with lowercase `vsm` or something like that.

NZJenkins replied:
`VertexShaderMode::` prefix is required because of `enum class` syntax

PatrickvL reviewed (changes requested):
A few questions / remarks, but VERY COOL to have this!!!

PatrickvL commented on src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShaderState.cpp:
```diff
@@ -0,0 +1,131 @@
+
+#else
+// HLSL
+#define arr(name, type, length) type name[length]
```
Somewhat silly, but when these two defines are swapped, defining `arr` last, it would mirror prettier with the undef's at the end.

NZJenkins replied:
iirc there was one `#undef` for both, which prevented mirroring - its now duplicated so both def and undef sections mirror

gandalfthewhite19890404 commented:
**Regression** - LEGO Star Wars 2 The original trilogy does not boot, hits test-case popup and crashes
LOG_TEST_CASE: Couldn't assemble vertex shader
In CompileHlsl (D:\a\Cxbx-Reloaded\Cxbx-Reloaded\src\core\hle\D3D8\Direct3D9\VertexShader.cpp line 266)

PatrickvL commented on src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShader.hlsl:
```diff
@@ -0,0 +1,600 @@
+	if (countFlag == 1)
+		texCoord.yzw = texCoord.x;
+	else if (countFlag == 2)
+		texCoord.zw = texCoord.y;
```
Should the source be `.yy`?

NZJenkins replied:
but that's extra typing!
https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-per-component-math
> f_4D.xzyw = pos.w; // write one component to more than one component

RadWolfie commented on projects/cxbx/CMakeLists.txt:
```diff
@@ -189,3 +189,18 @@ if(${CMAKE_GENERATOR} MATCHES "Visual Studio ([^9]|[9][0-9])")
+)
+
+# Copy shader files to hlsl directory
+add_custom_target(copy_hlsl ALL
```
Redundant code, see pull request #1932 for better solution using `file(COPY ...)` command. (1)

NZJenkins replied:
I assume you mean I can both deduplicate the hlsl copy code, and use `file(COPY ...)` with `${CMAKE_RUNTIME_OUTPUT_DIRECTORY}`?
Will look into it

RadWolfie replied:
Correct, see `projects/misc/batch.cmake` file from that pull request or in master branch. Basically your files you need to copy into build folder should be done inside batch.cmake file. You may not remember our previous discussion for this type of solution.

NZJenkins replied:
have redone the cmake part of the PR which should resolve this

RadWolfie commented on projects/cxbxr-emu/CMakeLists.txt:
```diff
@@ -162,3 +162,18 @@ target_link_libraries(cxbxr-emu
+)
+
+# Copy shader files to hlsl directory
+add_custom_target(copy_hlsl-emu ALL
```
Redundant code, see pull request #1932 for better solution using `file(COPY ...)` command. (2)

RadWolfie commented on projects/cxbx/CMakeLists.txt:
```diff
@@ -189,3 +189,18 @@ if(${CMAKE_GENERATOR} MATCHES "Visual Studio ([^9]|[9][0-9])")
+
+# HLSL compilation
+# Don't build HLSL files, though include them in Visual Studio
+set_source_files_properties(
```
IMHO, create a list in root CMakeLists.txt and refer it here will remove unnecessary redundant list. (1)

NZJenkins replied:
Yes good idea

RadWolfie commented on projects/cxbxr-emu/CMakeLists.txt:
```diff
@@ -162,3 +162,18 @@ target_link_libraries(cxbxr-emu
+
+# HLSL compilation
+# Don't build HLSL files, though include them in Visual Studio
+set_source_files_properties(
```
IMHO, create a list in root CMakeLists.txt and refer it here will remove unnecessary redundant list. (2)

RadWolfie commented on projects/cxbx/CMakeLists.txt:
```diff
@@ -189,3 +189,18 @@ if(${CMAKE_GENERATOR} MATCHES "Visual Studio ([^9]|[9][0-9])")
+set_source_files_properties(
+  "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/Direct3D9/CxbxVertexShaderTemplate.hlsl"
+  "${CXBXR_ROOT_DIR}/src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShader.hlsl" 
+  PROPERTIES VS_TOOL_OVERRIDE "None"
```
Had you tried with `HEADER_FILE_ONLY ON` instead of `VS_TOOL_OVERRIDE "None"`? If it works, then we could apply for the same with NV2A source files.

NZJenkins replied:
Yes `HEADER_FILE_ONLY` used to work, but VS seems to have started building it regardless...
Followed this answer suggesting `VS_TOOL_OVERRIDE` https://stackoverflow.com/a/46740948

NZJenkins replied:
I hit the issue again after switching between branches - it might be a Visual Studio + CMake issue.
Will restore `HEADER_FILE_ONLY` and see if I can get it to build again.

NZJenkins replied:
using HEADER_FILE_ONLY again

NZJenkins commented on src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShader.hlsl:
```diff
@@ -0,0 +1,600 @@
+        // else if TCI_OBJECT TODO is this just model position?
+		else if (tState.TexCoordIndexGen == TCI_SPHERE)
+		{
+            // TODO verify
```
Pretty sure this is verified... list test case/s and remove TODO

gandalfthewhite19890404 commented:
Artifacts does not includes hlsl folder, it is causing crashes with every titles, already discussed and verified on Discord.

NZJenkins commented:
Need to include `hlsl` in build artifacts
Other missing thing is, fixedfunction has no "passthrough" shader path! This was handled separately in #1894 (since Xbox internally uses a shader program rather than the fixed function backend.
Will look at patching in stopgap passthrough handling

RadWolfie commented:
@NZJenkins Some time in the future, I will have "install" functionality in CMake scripts into much generic usage. Since I have discover a way to do it with XbSymbolDatabase repo's CI action. Then every contiguous integration service doesn't need another update, **even per platform doesn't need extra work as well**. In fact, I should create a ticket for this feature.

PatrickvL commented:
This is why I was trying to avoid separate HLSL files

PatrickvL commented on src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShader.hlsl:
```diff
@@ -0,0 +1,616 @@
+}
+
+// Output registers
+struct VS_OUTPUT
```
For pixel shader HLSL, this exact output structure would need to be available as input, could this be moved into it's own file so it can be shared? Or shall that be postponed until the need arises?

Same for generic vertex shader HLSL btw...

PatrickvL commented on src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShader.hlsl:
```diff
@@ -0,0 +1,616 @@
+#undef CXBX_ALL_TEXCOORD_INPUTS // Enable this to disable semantics in VS_INPUT (instead, we'll use an array of generic TEXCOORD's)
+
+// Input registers
+struct VS_INPUT
```
Generic vertex shader HLSL probably overlaps with this. Could it be refactored a bit to avoid duplication?

gandalfthewhite19890404 commented:
What do you think about function rendering mode GUI indication?

NZJenkins commented:
> What do you think about function rendering mode GUI indication?

Some kind of overlay system would be great for visual feedback when pressing debug keys, but we don't have anything like that now.
If you mean a GUI setting to disable it - will look at this after other issues are addressed since F2 does the job for now.

medievil1 commented:
azurik(MS-007 V1.01) not happy with the current  state...lol 
![image](https://user-images.githubusercontent.com/38597905/97787701-457a1e80-1b8a-11eb-9ade-8fe8796335c2.png)

medievil1 commented:
azurik is better in game, start menu has the offset center picture though (issue popped up before) 
![image](https://user-images.githubusercontent.com/38597905/97804893-9fc5be80-1c20-11eb-8c55-fc544a52cbbe.png)

medievil1 commented:
and ummmm....lol
![image](https://user-images.githubusercontent.com/38597905/97805638-5f1c7400-1c25-11eb-8216-c5e08f69aebc.png)

PatrickvL commented on src/core/hle/D3D8/XbVertexShader.cpp:
```diff
@@ -1788,40 +1789,29 @@ void CxbxImpl_DeleteVertexShader(DWORD Handle)
+	float* constant_floats = HLE_get_NV2A_vertex_constant_float4_ptr(Register);
+	memcpy(constant_floats, pConstantData, ConstantCount * sizeof(float) * 4);
+
+	// Mark the constant as dirty, so that CxbxUpdateHostVertexShaderConstants will pick it up
+	extern NV2ADevice* g_NV2A; // TMP glue
+	auto nv2a = g_NV2A->GetDeviceState();
+	for (int i = 0; i < ConstantCount; i++) {
+		nv2a->pgraph.vsh_constants_dirty[Register + i] = true;
```
`CxbxUpdateHostVertexShaderConstants` doesn't check dirty anymore, so this could be removed or disabled..

NZJenkins replied:
yep left the dirty tracking in even though we're not reading it - wasn't worried about that. But now the comment is misleading so should probably fix one or the other

medievil1 commented:
azurik decided it all of a sudden in mid step wants to go black screen now.... 
![image](https://user-images.githubusercontent.com/38597905/98006855-bacf3480-1dc0-11eb-92bf-d9906332b370.png)

medievil1 commented:
KOTOR crashes after skipping video(immediately crashes) with two fatal errors 
`[0x2084] MAIN: Received Exception (Code := 0xC0000005)

 EIP := 0x658F82DA
 EFL := 0x00010206
 EAX := 0xE5AC3404 EBX := 0x00000000 ECX := 0x65A4AB20 EDX := 0x65ADCFE4
 ESI := 0x65AA983C EDI := 0x00000001 ESP := 0x0D73E5AC EBP := 0x0D73E5B0
 CR2 := 0x00000000

[0x2084] FATAL: X86     Popup : Received Exception Code 0xC0000005 @ EIP := 0x658F82DA

  Press "OK" to terminate emulation.
  Press "Cancel" to debug.
[0x31C8] MAIN: Received Exception (Code := 0xC0000005)
[0x3A94] MAIN: Received Exception (Code := 0xC0000005)

 EIP := 0x701B7BC3
 EFL := 0x00010297
 EAX := 0x3F800000 EBX := 0x65AC40A8 ECX := 0xCEC655EF EDX := 0x00000000
 ESI := 0x00000000 EDI := 0x701B7EB0 ESP := 0x1684FEB8 EBP := 0x1684FEDC
 CR2 := 0x00000000

[0x3A94] FATAL: X86     Popup : Received Exception Code 0xC0000005 @ EIP := 0x701B7BC3

  Press "OK" to terminate emulation.
  Press "Cancel" to debug.
`

NZJenkins commented:
@medievil1 I saw the Azurik issue happen, I suspect it's a random pre-existing issue. Did you try it more than once?
Fixed the KOTOR issue and cleaned up the fix commits.
Still some tricky regressions remaining

medievil1 commented:
> 
> 
> @medievil1 I saw the Azurik issue happen, I suspect it's a random pre-existing issue. Did you try it more than once?
> Fixed the KOTOR issue and cleaned up the fix commits.
> Still some tricky regressions remaining

yup, just tried again even on the latest and yea, it still goes black 
![image](https://user-images.githubusercontent.com/38597905/98140021-e589bd80-1e92-11eb-959b-33ac4b1333ed.png)

NZJenkins commented:
> yup, just tried again even on the latest and yea, it still goes black
> Of course the PR is 17 commits behind master so one or two of those may fix the issues...

It does the same thing on `master` [image](https://user-images.githubusercontent.com/9897874/98781857-e8b91680-245b-11eb-88f0-9278a9951f34.png)

jenci8888 commented:
Regression intro screen:
![image](https://user-images.githubusercontent.com/40495799/99159050-6047a980-26d9-11eb-956b-4c67bd770295.png)

I've checked against to master build (after 1894 PR)
![image](https://user-images.githubusercontent.com/40495799/99159064-853c1c80-26d9-11eb-89d8-808d815eaaa5.png)

NZJenkins commented:
Have pushed the latest work on this, fixing some regressions and addressing review comments - and rebased on `master`
Hopefully this is about done, functionally. Would be good to get some testing :)
Still some tidying to do

RadWolfie commented:
Waiting on pull request #2076 to be merge first. Then changes for include files will need to be revert and shift hlsl files to install command via `install(FILES ...)` usage.

medievil1 commented:
2076 has been merged :P looking forward to this getting merged too

NZJenkins commented:
Dropped CI globbing and added cmake install...

RadWolfie commented on CMakeLists.txt:
```diff
@@ -407,6 +412,12 @@ install(FILES ${cxbxr_INSTALL_files}
 )
 
+install(FILES
+	"${CMAKE_SOURCE_DIR}/src/core/hle/D3D8/Direct3D9/FixedFunctionVertexShaderState.hlsli"
```
No tabs anywhere in CMake relative files please. Use single space as indent.

RadWolfie commented:
LTGM for CMake relative changes. 👍

PatrickvL reviewed (approved):
Nzjenkins ftw!
